### ISSUE TITLE
Accessibility: support for screen readers

### ISSUE THREAD

**Issue by tronical** [slint-ui/slint#32]
At the risk of this being a very open task, this ticket tracks the foundation of accessibility support. This is by no means everything, but it's a start.

Acceptance Criteria: It's possible to navigate the widget gallery demo using voiceover and a keyboard on at least one platform (such as macOS or Linux). This means the text of labels, etc. is read, it's possible to use the keyboard to navigate between different elements and activate them.

**Comment by jacquetc** [comment]
Hello, 
AccessKit is still young but looks promising, I'm currently trying to understand it. I would not use 60fps for any desktop project because of the lack of accessibility.

Please, could you rename the title ? Something like "Accessibility: support for screen readers".

For your information, screen readers on Windows are: 
- Narrator (included in Windows)
- JAWs (free trial)
- NVDA (free), easy to launch

Linux (using AT-SPI 2):
- Orca 

Apple OSX & IPhone: 
- VoiceOver

Android:
- TalkBack

I found a useful [checklist](https://docs.flutter.dev/development/accessibility-and-localization/accessibility#accessibility-release-checklist) on [Flutter](https://docs.flutter.dev/development/accessibility-and-localization/accessibility).

**Comment by tronical** [comment]
Done :). Thanks for the feedback and the checklist!

**Comment by Jookia** [comment]
Just to add on this, no accessibility support makes Slint a really hard sell for a GUI framework compared to something like Qt, GTK or even stock Win32. Slint is barred from running in organizations that require use of accessible software for example.

**Comment by mwcampbell** [comment]
Since someone mentioned my AccessKit project, I'll offer a quick update on it. I've got a basic Windows implementation working; the biggest limitation right now is lack of support for text editing. Also, I have a [branch of Druid with AccessKit support](https://github.com/mwcampbell/druid/tree/accesskit). So now, the Slint team has something concrete to look at. I'll start working on AccessKit again in a few weeks.

**Comment by tronical** [comment]
That’s excellent. Thank you for the update (and your work on AccessKit)!

**Comment by DemiMarie** [comment]
I noticed this is listed as “low priority”.  Unfortunately, in multiple nations, inaccessible UIs in commercial products are prohibited by law, which makes this a blocker.  (Source: Hacker News)

**Comment by ogoffart** [comment]
@DemiMarie : Thanks for your comment.
We have actually started to work to improve our support for accessibility, and are actually working in an `accessibility` branch (#1193 and #1194).  But not something that can be tested yet.

**Comment by DemiMarie** [comment]
> @DemiMarie : Thanks for your comment. We have actually started to work to improve our support for accessibility, and are actually working in an `accessibility` branch (#1193 and #1194). But not something that can be tested yet.

You’re welcome!  I am glad that it was helpful, and even more so that you are working on accessibility support.

**Comment by DemiMarie** [comment]
@ogoffart: I hope this is not too much to ask, but could non-accessible elements cause a compiler warning?

**Comment by Be-ing** [comment]
I like that idea a lot.

**Comment by hunger** [comment]
Warning on all non-accessible elements seems a bit of overkill: You do not want the screen reader point out random Rectangles making up a Button! You care for the Button in its entirety and the Functionality that provides.

Maybe we could warn about non-accessible _exported_ Elements. Those are the building blocks used elsewhere.

**Comment by Be-ing** [comment]
Warning on all non-accessible elements might work okay if there's a way to explicitly opt out of the warning like clippy lints.

**Comment by DemiMarie** [comment]
> Maybe we could warn about non-accessible _exported_ Elements. Those are the building blocks used elsewhere.

That sounds like a great idea.

**Comment by mwcampbell** [comment]
Now that [AccessKit](https://github.com/AccessKit/accesskit) integration is merged in egui, and the AccessKit Windows backend supports text editing (coming soon on macOS), maybe it's a good time to look at integrating AccessKit into slint. That is, unless you expect that desktop apps in production will use the Qt backend for the long term, though that complicates commercial licensing.

**Comment by DemiMarie** [comment]
What about a direct AT-SPI backend for Linux?

**Comment by Be-ing** [comment]
That's what AccessKit is building: https://github.com/AccessKit/accesskit/issues/56

**Comment by mwcampbell** [comment]
If there are any concerns about AccessKit's implementation, API, or support for different widget types that are blocking Slint's adoption of AccessKit, please let me know. Thanks.

**Comment by tronical** [comment]
Beyond #2865 I'm going to file a few tickets under the [accessibility label](https://github.com/slint-ui/slint/labels/accessibility), so that we can track further improvements. My objective is to close this task when #2865 lands and incrementally work forwards.
